File: wiki_java/src/main/java/com/github/klee0kai/wiki/provide/identifiers/ThreadsModule.java
package com.github.klee0kai.wiki.provide.identifiers;

import com.github.klee0kai.stone.annotations.module.Module;
import com.github.klee0kai.stone.annotations.module.Provide;

import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

@Module
public class ThreadsModule {

    @ThreadQualifier(type = ThreadQualifier.ThreadType.Main)
    @Provide(cache = Provide.CacheType.Strong)
    public ThreadPoolExecutor mainThreadPoolExecutor() {
        return new ThreadPoolExecutor(0, 1, 0, TimeUnit.MILLISECONDS, new LinkedBlockingDeque<>(), runnable -> new Thread(runnable));
    }

}

File: wiki_kotlin/src/main/kotlin/com/github/klee0kai/wiki/provide/identifiers/ThreadsModule.kt
package com.github.klee0kai.wiki.provide.identifiers

import com.github.klee0kai.stone.annotations.module.Module
import com.github.klee0kai.stone.annotations.module.Provide
import java.util.concurrent.LinkedBlockingDeque
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

@Module
open class ThreadsModule {

    @ThreadQualifier(type = ThreadQualifier.ThreadType.Main)
    @Provide(cache = Provide.CacheType.Strong)
    open fun mainThreadPoolExecutor(): ThreadPoolExecutor {
        return ThreadPoolExecutor(
            0,
            1,
            0,
            TimeUnit.MILLISECONDS,
            LinkedBlockingDeque()
        ) { runnable: Runnable? -> Thread(runnable) }
    }

}